File: OsmAnd/src/net/osmand/plus/NotificationHelper.java
package net.osmand.plus;

import android.annotation.TargetApi;
import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.content.Context;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.NotificationCompat.Builder;
import androidx.core.app.NotificationManagerCompat;

import net.osmand.plus.notifications.DownloadNotification;
import net.osmand.plus.notifications.ErrorNotification;
import net.osmand.plus.notifications.GpxNotification;
import net.osmand.plus.notifications.NavigationNotification;
import net.osmand.plus.notifications.OsmandNotification;
import net.osmand.plus.notifications.OsmandNotification.NotificationType;

import java.util.ArrayList;
import java.util.List;

public class NotificationHelper {

	public static final String NOTIFICATION_CHANEL_ID = "osmand_background_service";
	private OsmandApplication app;

	private NavigationNotification navigationNotification;
	private GpxNotification gpxNotification;
	private DownloadNotification downloadNotification;
	private ErrorNotification errorNotification;
	private List<OsmandNotification> all = new ArrayList<>();

	public NotificationHelper(OsmandApplication app) {
		this.app = app;
		init();
	}

	private void init() {
		navigationNotification = new NavigationNotification(app);
		gpxNotification = new GpxNotification(app);
		downloadNotification = new DownloadNotification(app);
		errorNotification = new ErrorNotification(app);
		all.add(navigationNotification);
		all.add(gpxNotification);
		all.add(downloadNotification);
	}

	public Notification buildTopNotification() {
		OsmandNotification notification = acquireTopNotification();
		if (notification != null) {
			removeNotification(notification.getType());
			setTopNotification(notification);
			Builder notificationBuilder = notification.buildNotification(false);
			if (notificationBuilder != null) {
				return notificationBuilder.build();
			} else {
				return null;
			}
		} else {
			return null;
		}
	}

	@NonNull
	public Notification buildDownloadNotification() {
		return downloadNotification.buildNotification(false).build();
	}

	public Notification buildErrorNotification() {
		removeNotification(errorNotification.getType());
		setTopNotification(errorNotification);
		return errorNotification.buildNotification(false).build();
	}

	@Nullable
	private OsmandNotification acquireTopNotification() {
		if (navigationNotification.isEnabled()) {
			return navigationNotification;
		} else if (gpxNotification.isEnabled() && gpxNotification.isActive()) {
			return gpxNotification;
		} else {
			return null;
		}
	}

	public void updateTopNotification() {
		OsmandNotification notification = acquireTopNotification();
		setTopNotification(notification);
	}

	private void setTopNotification(OsmandNotification notification) {
		for (OsmandNotification n : all) {
			n.setTop(n == notification);
		}
	}

	public void showNotifications() {
		if (!hasAnyTopNotification()) {
			removeTopNotification();
		}
		for (OsmandNotification notification : all) {
			notification.showNotification();
		}
	}

	public void refreshNotification(NotificationType notificationType) {
		for (OsmandNotification notification : all) {
			if (notification.getType() == notificationType) {
				notification.refreshNotification();
				break;
			}
		}
	}

	public void onNotificationDismissed(NotificationType notificationType) {
		for (OsmandNotification notification : all) {
			if (notification.getType() == notificationType) {
				notification.onNotificationDismissed();
				break;
			}
		}
	}

	public boolean hasAnyTopNotification() {
		for (OsmandNotification notification : all) {
			if (notification.isTop()) {
				return true;
			}
		}
		return false;
	}

	public void refreshNotifications() {
		if (!hasAnyTopNotification()) {
			removeTopNotification();
		}
		for (OsmandNotification notification : all) {
			notification.refreshNotification();
		}
	}

	public void removeTopNotification() {
		NotificationManagerCompat notificationManager = NotificationManagerCompat.from(app);
		notificationManager.cancel(OsmandNotification.TOP_NOTIFICATION_SERVICE_ID);
	}

	public void removeNotification(NotificationType notificationType) {
		for (OsmandNotification notification : all) {
			if (notification.getType() == notificationType) {
				notification.removeNotification();
				break;
			}
		}
	}

	public void removeNotifications(boolean inactiveOnly) {
		for (OsmandNotification notification : all) {
			if (!inactiveOnly || !notification.isActive()) {
				notification.removeNotification();
			}
		}
	}

	@TargetApi(26)
	public void createNotificationChannel() {
		if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
			NotificationChannel channel = new NotificationChannel(NOTIFICATION_CHANEL_ID,
					app.getString(R.string.osmand_service), NotificationManager.IMPORTANCE_LOW);
			channel.enableVibration(false);
			channel.setDescription(app.getString(R.string.osmand_service_descr));
			NotificationManager mNotificationManager = (NotificationManager) app
					.getSystemService(Context.NOTIFICATION_SERVICE);
			mNotificationManager.createNotificationChannel(channel);
		}
	}
}

File: OsmAnd-telegram/src/main/java/net/osmand/telegram/notifications/NotificationHelper.kt
package net.osmand.telegram.notifications

import android.annotation.TargetApi
import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.content.Context
import android.support.v4.app.NotificationManagerCompat
import net.osmand.telegram.R
import net.osmand.telegram.TelegramApplication
import net.osmand.telegram.notifications.TelegramNotification.NotificationType
import java.util.*

class NotificationHelper(private val app: TelegramApplication) {

    private var shareLocationNotification: ShareLocationNotification? = null
    private val all = ArrayList<TelegramNotification>()

    init {
        init()
    }

    private fun init() {
        val shareLocationNotification = ShareLocationNotification(app)
        this.shareLocationNotification = shareLocationNotification
        all.add(shareLocationNotification)
    }

    fun buildTopNotification(): Notification? {
        val notification = acquireTopNotification()
        if (notification != null) {
            removeNotification(notification.type)
            setTopNotification(notification)
            val notificationBuilder = notification.buildNotification(false)
            return notificationBuilder?.build()
        }
        return null
    }

    private fun acquireTopNotification(): TelegramNotification? {
        var notification: TelegramNotification? = null
        if (shareLocationNotification!!.isEnabled && shareLocationNotification!!.isActive) {
            notification = shareLocationNotification
        }
        return notification
    }

    fun updateTopNotification() {
        val notification = acquireTopNotification()
        setTopNotification(notification)
    }

    private fun setTopNotification(notification: TelegramNotification?) {
        for (n in all) {
            n.isTop = n === notification
        }
    }

    fun showNotifications() {
        if (!hasAnyTopNotification()) {
            removeTopNotification()
        }
        for (notification in all) {
            notification.showNotification()
        }
    }

    fun refreshNotification(notificationType: NotificationType) {
        for (notification in all) {
            if (notification.type == notificationType) {
                notification.refreshNotification()
                break
            }
        }
    }

    fun onNotificationDismissed(notificationType: NotificationType) {
        for (notification in all) {
            if (notification.type == notificationType) {
                notification.onNotificationDismissed()
                break
            }
        }
    }

    fun hasAnyTopNotification(): Boolean {
        for (notification in all) {
            if (notification.isTop) {
                return true
            }
        }
        return false
    }

    fun refreshNotifications() {
        if (!hasAnyTopNotification()) {
            removeTopNotification()
        }
        for (notification in all) {
            notification.refreshNotification()
        }
    }

    fun removeTopNotification() {
        val notificationManager = NotificationManagerCompat.from(app)
        notificationManager.cancel(TelegramNotification.TOP_NOTIFICATION_SERVICE_ID)
    }

    fun removeNotification(notificationType: NotificationType) {
        for (notification in all) {
            if (notification.type == notificationType) {
                notification.removeNotification()
                break
            }
        }
    }

    fun removeNotifications() {
        for (notification in all) {
            notification.removeNotification()
        }
    }

    @TargetApi(26)
    fun createNotificationChannel() {
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.O) {
            val channel = NotificationChannel(NOTIFICATION_CHANEL_ID,
                    app.getString(R.string.osmand_service), NotificationManager.IMPORTANCE_LOW)
            channel.enableVibration(false)
            channel.description = app.getString(R.string.osmand_service_descr)
            val mNotificationManager = app
                    .getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            mNotificationManager.createNotificationChannel(channel)
        }
    }

    companion object {
        const val NOTIFICATION_CHANEL_ID = "osmand_telegram_background_service"
    }
}
